implemented shopping cart

ADIGrimm opened the pull request.

ADIGrimm committed:
implemented shopping cart

Elena-Bruyako commented on src/main/java/online/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,41 @@
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @MapsId
+    @JoinColumn(name = "user_id")
```
```suggestion
    @JoinColumn(name = "id", nullable = false)
```

Elena-Bruyako commented on src/main/java/online/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,41 @@
+@Table(name = "carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/ShoppingCartService.java:
```diff
@@ -0,0 +1,19 @@
+            UpdateCartItemQuantityDto updateQuantityDto
+    );
+
+    ShoppingCartDto deleteBook(Long cartItemId, Long userId);
```
```suggestion
    void deleteBook(Long cartItemId, Long userId);
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    @Transactional
+    public ShoppingCartDto addBook(Long id, AddCartItemDto addCartItemDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id).orElseThrow(()
```
Find book by id first
check if book already exists in cart, if yes - increase quantity, if no - add to cart

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+                -> new EntityNotFoundException("Can't find shopping cart by id " + id)));
+    }
+
+    @Override
```
```suggestion
    @Override
    @Transactional
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+    ) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(userId).orElseThrow(()
+                -> new EntityNotFoundException("Can't find shopping cart by id " + userId));
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
```
```suggestion
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, shoppingCart.getId()).....
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
```
```suggestion
    @Override
    @Transactional
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+    public ShoppingCartDto deleteBook(Long cartItemId, Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(userId).orElseThrow(()
+                -> new EntityNotFoundException("Can't find shopping cart by id " + userId));
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
```
```suggestion
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId())....
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBook(Long id, AddCartItemDto addCartItemDto) {
```
```suggestion
    public ShoppingCartDto addBook(Long userId, AddCartItemDto addCartItemDto) {
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Override
+    public ShoppingCartDto getInfo(Long id) {
```
```suggestion
    public ShoppingCartDto getInfo(Long userId) {
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/UserServiceImpl.java:
```diff
@@ -36,6 +39,9 @@ public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         user.setRoles(Set.of(roleRepository.findByName(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Can't find role"))));
+        ShoppingCart cart = new ShoppingCart();
+        cart.setUser(user);
+        shoppingCartRepository.save(cart);
         return userMapper.toDto(userRepository.save(user));
```
create method in shoppingCartService `createShoppingCart` and implement it 

```suggestion
        userRepository.save(user)
        shoppingCartService.createShoppingCart(user);
        return userMapper.toDto(user);
```

Elena-Bruyako reviewed (changes requested):
see comments

Elena-Bruyako commented on src/main/java/online/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,73 @@
+            @Valid @RequestBody AddCartItemDto addCartItemDto,
+            Authentication authentication
+    ) {
+        User user = (User) authentication.getPrincipal();
```
You can create separate method to avoid duplication of code to get `Long userId `

Elena-Bruyako commented on src/main/java/online/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,73 @@
+            description = "Delete book from cart and return cart info")
+    @DeleteMapping("/items/{cartItemId}")
+    @PreAuthorize("hasAnyRole('ROLE_USER')")
+    public ShoppingCartDto deleteBook(
```
```suggestion
    public void deleteBook(
```

Elena-Bruyako commented on src/main/java/online/bookstore/dto/cart/AddCartItemDto.java:
```diff
@@ -0,0 +1,13 @@
+
+@Data
+public class AddCartItemDto {
+    @NotBlank
```
Use `@NotBlank` for String

```suggestion
    @NotNull
    @Positive
```

Elena-Bruyako commented on src/main/java/online/bookstore/dto/cart/AddCartItemDto.java:
```diff
@@ -0,0 +1,13 @@
+public class AddCartItemDto {
+    @NotBlank
+    private Long bookId;
+    @Min(1)
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/online/bookstore/dto/cart/UpdateCartItemQuantityDto.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+public class UpdateCartItemQuantityDto {
+    @Min(1)
```
```suggestion
    @Positive
```

Elena-Bruyako commented on src/main/java/online/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @NotNull
+    @Min(value = 1)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/online/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/online/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne(fetch = FetchType.LAZY)
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/online/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,41 @@
+
+@Entity
+@SQLDelete(sql = "UPDATE carts SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction("is_deleted = FALSE")
```
add `private boolean isDeleted;`

Elena-Bruyako commented on src/main/java/online/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,41 @@
+    @OneToOne(fetch = FetchType.LAZY)
+    @MapsId
+    @JoinColumn(name = "user_id")
+    @NotNull
```
```suggestion
```

ADIGrimm committed:
fixed

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    @Transactional
```
```suggestion
```

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/UserServiceImpl.java:
```diff
@@ -36,6 +38,8 @@ public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
         user.setRoles(Set.of(roleRepository.findByName(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Can't find role"))));
-        return userMapper.toDto(userRepository.save(user));
+        user = userRepository.save(user);
```
```suggestion
        userRepository.save(user);
```

Elena-Bruyako reviewed (changes requested):
see comments

ADIGrimm committed:
fixed

removed transactional annotation from createShoppingCart()
removed override of user variable in registration()

ADIGrimm committed:
fixed to return testing capability

(error "field id/user_id doesn't have a default value" was fixed)
added generated value to ShoppingCart id field
changed user id column name
